### PR TITLE
[web] Refactor to do all justify alignment during layout

### PR DESCRIPTION
**Pull request by mdebbar**
Before this PR: The `justify` alignment occurs during the painting of text on canvas2d.
After this PR: Do all the `justify` alignment calculation during layout.

Benefits:
- We get the same alignment on both canvas2d & DOM for free! (especially after https://github.com/flutter/engine/pull/31907).
- Selection boxes and placeholders are positioned correctly.

To achieve this, we have to move the box-positioning logic out of the `LineBuilder`. It has to happen after all the lines have been constructed so that `justify` alignment can be calculated.

Fixes https://github.com/flutter/flutter/issues/96533

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 6.
View them at https://flutter-engine-gold.skia.org/cl/github/31937

File: lib/web_ui/lib/src/engine/text/layout_service.dart [hunk 1]
```diff
@@ -270,6 +287,145 @@ class TextLayoutService {
...
+  /// Calculates for the given [line], the amount of extra width that needs to be
+  /// added to each space box in order to align the line with the rest of the
+  /// paragraph.
+  double _calculateJustifyPerSpaceBox(EngineLineMetrics line) {
+    final double justifyTotal = paragraph.width - line.width;
+    final RangeBox lastBox = line.boxes.last;
+
+    int spaceBoxesToJustify = line.spaceBoxCount;
+    // If the last box is a space box, we can't use it to justify text.
+    if (lastBox is SpanBox && lastBox.isSpaceOnly) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text/layout_service.dart:419]: Is it guaranteed that there's no more than one space-only box?

File: lib/web_ui/lib/src/engine/text/layout_service.dart [hunk 2]
```diff
@@ -270,6 +287,145 @@ class TextLayoutService {
...
+    int i = 0;
+    double cumulativeWidth = 0.0;
+    while (i < boxes.length) {
+      final RangeBox box = boxes[i];
+      if (box.boxDirection == _paragraphDirection) {
+        // The box is in the same direction as the paragraph.
+        box.startOffset = cumulativeWidth;
+        box.lineWidth = line.width;
+        if (box is SpanBox && box.isSpaceOnly) {
+          box._width += justifyPerSpaceBox;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text/layout_service.dart:312]: Do we need to skip increasing the width of the last space-only box? Or maybe the last space-only box should have zero width if the text is justified? :thinking:

File: lib/web_ui/lib/src/engine/text/layout_service.dart [hunk 3]
```diff
@@ -270,6 +287,145 @@ class TextLayoutService {
...
+    final List<RangeBox> boxes = line.boxes;
+    double cumulativeWidth = 0.0;
+    for (int i = last; i >= first; i--) {
+      // Update the visual position of each box.
+      final RangeBox box = boxes[i];
+      assert(box.boxDirection != _paragraphDirection);
+      box.startOffset = startOffset + cumulativeWidth;
+      box.lineWidth = line.width;
+      if (box is SpanBox && box.isSpaceOnly) {
+        box._width += justifyPerSpaceBox;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text/layout_service.dart:402]: ditto w.r.t. last space-only box

**Review by yjbanov — APPROVED** [review]

File: lib/web_ui/lib/src/engine/text/layout_service.dart (hunk 2 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/layout_service.dart:312]: Good intuition! I was going by what the browser does. But now I checked Flutter, and you're right, it collapses the width of trailing spaces!

Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/layout_service.dart:312]: FWIW, Flutter always collapses trailing spaces (regardless of which alignment is used).

Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/layout_service.dart:312]: Actually, I (partially) lied.

During text painting/rendering, Flutter treats trailing spaces as if they were zero-width (e.g. no background).

During text editing (cursor placement & selection highlighting), Flutter treats trailing spaces the same as other spaces.

| Left | Right | Justify |
|-|-|-|
| image | image | image |

I believe what's happening during text editing is they have no special case for trailing spaces, but they clamp the results of `getBoxesForRange` to the width of the paragraph.

File: lib/web_ui/lib/src/engine/text/layout_service.dart (hunk 3 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/layout_service.dart:402]: Done.

File: lib/web_ui/lib/src/engine/text/layout_service.dart (hunk 2 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/layout_service.dart:312]: Done (minus the clamping part).

File: lib/web_ui/lib/src/engine/text/layout_service.dart (hunk 1 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/layout_service.dart]: Done.

File: lib/web_ui/lib/src/engine/text/layout_service.dart (hunk 2 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/layout_service.dart:312]: Filed an issue to implement the clamping: https://github.com/flutter/flutter/issues/100567

**Pull request merged**